File: source/Transition/Transition.java
package source.Transition;

import com.badlogic.gdx.graphics.glutils.FrameBuffer;
import com.badlogic.gdx.graphics.g2d.Batch;
import com.badlogic.gdx.graphics.g2d.TextureRegion;
import com.badlogic.gdx.graphics.Texture;
import source.State.StateManager;

/**
 * @author Xico
 */

public interface Transition {

    int WORLD_WIDTH = StateManager.getInstance().getWorldWidth();
    int WORLD_HEIGHT = StateManager.getInstance().getWorldHeight();

    /**
     * Starts the transition.
     */
    void start();

    /**
     * @return {@code true} if the transition is running, {@code false} otherwise.
     */
    boolean isRunning();

    /**
     * @return {@code true} if the transition has finished, {@code false} otherwise.
     */
    boolean isCompleted();

    /**
     * Finishes the transition.
     */
    void finish();

    /**
     * Updates the transition.
     *
     * @param delta - time elapsed since the last call to update.
     */
    void update(float delta);

    /**
     * Renders the current and next textures to the given {@link Batch}.
     * These textures were obtained using the Render-To-Texture (RTT) technique.
     * A {@link TextureRegion} is used instead of a {@link Texture} because the framebuffer is upside
     * down to normal textures, ...
     * ...as seen here: <a href="http://www.gamefromscratch.com/post/2014/01/21/Rendering-a-3D-model-to-texture-in-LibGDX.aspx"></a>
     *
     * @param batch   {@link Batch} used to render.
     * @param current current (or previous?) screen's {@link TextureRegion}, rendered onto a {@link FrameBuffer}.
     * @param next    next screen's {@link TextureRegion}, rendered onto a {@link FrameBuffer}.
     * @see <a href="http://apprize.info/programming/libgdx/10.html">Transitions</a>
     */
    void render(Batch batch, TextureRegion current, TextureRegion next);

}

File: core/src/main/kotlin/cyberpunk/core/transition/Transition.kt
package cyberpunk.core.transition

import com.badlogic.gdx.graphics.g2d.Batch
import com.badlogic.gdx.graphics.g2d.TextureRegion

interface Transition {

  /**
   * Abstract property - override in concrete implementations.
   */
  var running: Boolean

  /**
   * Starts the transition.
   */
  fun start() {
    running = true
  }

  /**
   * Finishes the transition.
   */
  fun finish() {
    running = false
  }

  /**
   * Checks whether or not this [Transition] is running.
   * @return true if it is running, false otherwise.
   */
  fun running() = running

  /**
   * Checks whether or not this [Transition] is completed.
   * @return true if it is completed, false otherwise.
   */
  fun completed() = !running

  /**
   * Updates the [Transition].
   * @param delta - time elapsed since the last call to update.
   */
  fun update(delta: Float)

  /**
   * Renders the current and next [TextureRegion]s to the given [Batch].
   * These textures were obtained using the Render-To-Texture (RTT) technique.
   * A [TextureRegion] is used instead of a texture because the framebuffer is upside down to normal textures,
   * as seen here: "http://www.gamefromscratch.com/post/2014/01/21/Rendering-a-3D-model-to-texture-in-LibGDX.aspx
   * @param batch   [Batch] used to render.
   * @param current current (or previous?) screen's [TextureRegion], rendered onto a frame buffer.
   * @param next    next screen's [TextureRegion], rendered onto a frame buffer.
   * @see "http://apprize.info/programming/libgdx/10.html"
   */
  fun render(batch: Batch, current: TextureRegion, next: TextureRegion)

}